File: villager/src/main/java/br/com/zup/charlescd/villager/infrastructure/persistence/ModuleRepository.java
/*
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
 * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package br.com.zup.charlescd.villager.infrastructure.persistence;

import io.quarkus.hibernate.orm.panache.PanacheRepository;

import javax.enterprise.context.ApplicationScoped;
import java.util.List;
import java.util.Optional;

@ApplicationScoped
public class ModuleRepository implements PanacheRepository<ModuleEntity> {

    public List<ModuleEntity> findByJobId(String ciJobId) {
        return find("ciJobId", ciJobId).list();
    }

    public Optional<ModuleEntity> findByNameAndTagNameAndStatus(String name, String tagName, ModuleBuildStatus status) {
        return Optional.ofNullable(find("name = ?1 and tagName = ?2 and status = ?3", name, tagName, status).firstResult());
    }

    public List<ModuleEntity> findModulesByStatus(ModuleBuildStatus status) {
        return find("status = ?1", status).list();
    }

    public List<ModuleEntity> findByBuildId(String buildId) {
        return find("build_id = ?1", buildId).list();
    }

    public void persist(List<ModuleEntity> moduleEntityList) {
        persist(moduleEntityList.stream());
    }

}

File: moove/darwin-repository/src/main/kotlin/br/com/zup/darwin/repository/ModuleRepository.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package br.com.zup.darwin.repository

import br.com.zup.darwin.entity.Module
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import java.util.*

interface ModuleRepository : JpaRepository<Module, String> {

    fun findByIdAndApplicationId(id: String, applicationId: String): Optional<Module>

    fun findAllByApplicationId(applicationId: String, pageable: Pageable): Page<Module>

    fun findByNameAndApplicationIdIgnoreCaseContaining(
        name: String,
        applicationId: String,
        page: Pageable
    ): Page<Module>

    fun findAllByIdAndApplicationId(ids: List<String>, applicationId: String): List<Module>

    @Query(nativeQuery = true, value = "SELECT DISTINCT (m.*) FROM modules m " +
            " INNER JOIN features_modules fm ON fm.module_id  = m.id " +
            " INNER JOIN features f ON f.id  = fm.feature_id  " +
            " INNER JOIN builds_features bf ON bf.feature_id  = f.id " +
            " INNER JOIN builds b ON b.id = bf.build_id " +
            " INNER JOIN deployments d ON d.build_id = b.id " +
            " WHERE d.circle_id = :circleId ")
    fun findAllModulesDeployedAtCircle(circleId: String): Optional<List<Module>>

}
